transaction/ResourceId.hpp

```diff
@@ -0,0 +1,217 @@
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_TRANSACTION_RESOURCE_ID_HPP_
+#define QUICKSTEP_TRANSACTION_RESOURCE_ID_HPP_
+
+#include <limits>
+#include <string>
+#include <type_traits>
+
+#include "catalog/CatalogTypedefs.hpp"
+#include "storage/StorageBlockInfo.hpp"
+#include "utility/HashPair.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+namespace transaction {
+
+/** \addtogroup Transaction
+ *  @{
+ */
+
+/**
+ * @brief Class for representing unique id for each database resource
+ *        e.g Tuple, block, relation, database etc...
+ **/
+class ResourceId {
+ public:
+  static constexpr database_id kDbIdPlaceholder =
+    std::numeric_limits<database_id>::max();
+  static constexpr relation_id kRelIdPlaceholder =
+    std::numeric_limits<relation_id>::max();
+  static constexpr block_id kBlockIdPlaceholder =
+    std::numeric_limits<block_id>::max();
+  static constexpr tuple_id kTupleIdPlaceholder =
+    std::numeric_limits<tuple_id>::max();
+
+  /**
+   * @brief Constructor for invalid resource id.
+   **/
+  ResourceId()
+    : ResourceId(kDbIdPlaceholder,
+                 kRelIdPlaceholder,
+                 kBlockIdPlaceholder,
+                 kTupleIdPlaceholder) {
+  }
+
+  /**
+   * @brief Constructor for database level resource id.
+   *
+   * @param db_id unique id of database resource.
+   **/
+  explicit ResourceId(database_id db_id)
+    : ResourceId(db_id,
+                 kRelIdPlaceholder,
+                 kBlockIdPlaceholder,
+                 kTupleIdPlaceholder) {
+  }
+
+  /**
+   * @brief Constructor for relation level resource id.
+   *
+   * @param db_id unique id of database resource.
+   * @param rel_id unique id of relation resource in database.
+   **/
+  ResourceId(database_id db_id, relation_id rel_id)
+    : ResourceId(db_id,
+                 rel_id,
+                 kBlockIdPlaceholder,
+                 kTupleIdPlaceholder) {
+  }
+
+  /**
+   * @brief Constructor for block level resource id.
+   *
+   * @param db_id unique id of the databse resource.
+   * @param rel_id unique id of the relation resource in the database.
+   * @param block_id unique id of the block resource in the relation.
+   **/
+  ResourceId(database_id db_id, relation_id rel_id, block_id block_id)
+    : ResourceId(db_id,
+                 rel_id,
+                 block_id,
+                 kTupleIdPlaceholder) {
+  }
+
+  /**
+   * @brief Constructor for tuple level resource id.
+   *
+   * @param db_id Unique id of the database resource.
+   * @param rel_id Unique id of the relation resource in the database.
+   * @param block_id Unique id of the block resource in the relation.
+   * @param tuple_id Unique id of the tuple resource in the block.
+   **/
+  ResourceId(database_id db_id,
+             relation_id rel_id,
+             block_id block_id,
+             tuple_id tuple_id)
```

Comment:
Mark all four arguments `const`.

Why not give these arguments the default placeholder values?